pybossa/jobs.py

```diff
@@ -622,21 +622,99 @@ def send_mail(message_dict, mail_all=False):
+        ''')
+    params = dict(project_id=project_id)
+    total_result = db.session.execute(sql, params).scalar()
+
+
+    sql = text('''
+        SELECT COUNT(*) FROM counter WHERE
+        project_id=:project_id;
+        ''')
+    params = dict(project_id=project_id)
+    total_counter = db.session.execute(sql, params).scalar()
+    return total_tasks, total_taskrun, total_result, total_counter
+
+
+def cleanup_task_records(project_id, limit):
+    """Cleanup records associated with task from all related tables."""
+
+    from sqlalchemy.sql import text
+    from pybossa.core import db
+
+    tables = ["counter", "result", "task_run", "task"]
+    task_ids = []
+    params = {"project_id": project_id, "limit": limit}
+    sql = text('''
+        SELECT id FROM task WHERE project_id=:project_id LIMIT :limit;
+        ''')
+    response = db.session.execute(sql, params).fetchall()
+    task_ids = [id[0] for id in response]
+    if not task_ids:
+        return
+
+    current_app.logger.info("Task ids staged for deletion: %s", task_ids)
+    task_ids_tuple = tuple(task_ids)
+    for table in tables:
+        sql = f"DELETE FROM {table} "
+        sql += "WHERE id IN :taskids;" if table == "task" else "WHERE task_id IN :taskids;"
+        db.session.execute(sql, {"taskids": task_ids_tuple})
+        db.session.commit()
+    current_app.logger.info("Task ids staged deleted from tables %s", tables)
+
+
+def delete_bulk_tasks_in_batches(project_id, force_reset):
+    """Delete bulk tasks in batches from project."""
+
+    batch_size = 100
+    total_task, total_taskrun, total_result, total_counter = count_rows_to_delete(project_id)
+    current_app.logger.info("total records to delete. task %d, task_run %d, result %d, counter %d", 
+                            total_task, total_taskrun, total_result, total_counter)
+
+    # count iterations required to delete records from all tables
+    counter_iterations = int(total_counter / batch_size) + (1 if total_counter % batch_size else 0)
+    result_iterations = int(total_result / batch_size) + (1 if total_result % batch_size else 0)
+    taskrun_iterations = int(total_taskrun / batch_size) + (1 if total_taskrun % batch_size else 0)
+    task_iterations = int(total_task / batch_size) + (1 if total_task % batch_size else 0)
+    current_app.logger.info("total iterations. task %d, task_run %d, result %d, counter %d", 
+                            task_iterations, taskrun_iterations, result_iterations, counter_iterations)
+
+    # TODO: implement force_reset = False
+    limit = batch_size
+    total_iterations = max(counter_iterations, result_iterations, taskrun_iterations, task_iterations)
+    for i in range(total_iterations):
+        count_deleted_records = i * batch_size
+        current_app.logger.info("Count of records deleted: %d", count_deleted_records)
+        cleanup_task_records(project_id=project_id, limit=limit)
+        time.sleep(2) # allow sql queries other than delete records to execute
```

Comment:
Consider moving the delay time to a configurable parameter, either defined at the top of the file or in settings_local.